### ISSUE TITLE
still able to communicate on channel after bad refresh

### ISSUE THREAD

**Issue by lmj0011** [centrifugal/centrifugo#99]
Does a channel not disconnect automatically when the refreshEndpoint returns error status?
I am still able to communicate in the channel that I'm connect to even after the 401 error returned from the refreshEndpoint

![not_connected](https://cloud.githubusercontent.com/assets/9396189/17257490/f6dc12fa-5587-11e6-8f97-523670349af0.png)

also is there a way to read the response from the refreshEndpoint, from the clientside?
https://fzambia.gitbooks.io/centrifugal/content/clients/javascript.html#configuration-parameters

**Comment by FZambia** [comment]
> Does a channel not disconnect automatically when the refreshEndpoint returns error status?

Currently it's implemented in a way that channels continue to be available for a client after error refresh response. This was done to give client a window to rescue its connection from being closed. So after 3 seconds it will try to refresh credentials again. And then after some time (`expired_connection_close_delay` option in configuration, by default 25 seconds) Centrifugo will close its connection automatically.

I have not used this connection lifetime feature myself (and I doubt anyone else used it, so you are first maybe). So I just reproduced your case - looks like it works like it should but some tweaks should maybe be made - I mean should client try to reconnect or not after this, should it continue to refresh credentials or not. What is your thoughts about this - what you expect to happen with such client after its credentials expired?

> is there a way to read the response from the refreshEndpoint, from the clientside

Why do you want this?

**Comment by FZambia** [comment]
Ah you have written why you need refresh response callback in gitter chat - so you want to notify user about this. It's not obvious for me why you want this. What you want to write in this alert? Why user should know about this? Again this feature is not fully investigated - so any feedback is valuable to make it comfortable to use.

**Comment by lmj0011** [comment]
>Why do you want this?

In my app I'm passing the request for the refresh endpoint through middleware that verifies if the user is still logged into the app. If they are not, I need to be able to read the response and alert them that their chat session has ended

**Comment by lmj0011** [comment]
>I mean should client try to reconnect or not after this, should it continue to refresh credentials or not. What is your thoughts about this - what you expect to happen with such client after its credentials expired (I mean after that window to rescue passed and Centrifugo finally disconnected client)?

I think that the disconnect event should fire after the refresh endpoint fails, and it shouldn't try to refresh until the client establishes a new connection. But this method would not be beneficial for all, so maybe there should be an refresh_attempts option. Where after a set number of refreshes centrifugo cuts off connection to the client and fires some event to signify that.

**Comment by FZambia** [comment]
So added 2 new options to javascript client in [3bd82a](https://github.com/centrifugal/centrifuge-js/commit/3bd82a379389351585128b7f20db36b2b704a34d) - `refreshAttempts` and `onSessionExpired`:
```
        // initialize Centrifuge object
        var centrifuge = new Centrifuge({
            ...
            refreshAttempts: 1,
            onSessionExpired: function() {
                console.log("Dear user, your credentials expired, please refresh your page");
            }
        });
```
By default everything works like before - i.e. unlimited refresh attempts and no callback. I've also built js files in master branch - could you test it? If ok we can make new release.

**Comment by FZambia** [comment]
Just renamed `onSessionExpired` to `refreshFailed` to be more semantically correct - as session could be actually expired before all refresh attempts will be used.

**Comment by lmj0011** [comment]
>could you test it?

**Comment by lmj0011** [comment]
verified after 1 failed refreshAttempt, the refreshFailed callback was called and no more refresh attempts were made after that. :+1: 

```javascript
let centrifuge = new Centrifuge({

           //[...]
            refreshAttempts: 1,
            refreshFailed: function() {
                bootbox.alert("Your chat session has expired, please refresh your page");
            }
        });
```
<br>
![refresh_attempt_1](https://cloud.githubusercontent.com/assets/9396189/17277749/758fb62c-5710-11e6-938a-691f8db1d478.png)

<br>
![chat_session_expired](https://cloud.githubusercontent.com/assets/9396189/17277750/7b61ebce-5710-11e6-97b5-369825fcf8d2.png)

**Comment by FZambia** [comment]
Released centrifuge-js 1.3.6
